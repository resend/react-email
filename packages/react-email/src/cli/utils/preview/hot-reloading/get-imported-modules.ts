import { parse } from '@babel/parser';
import traverse from '@babel/traverse';

export const getImportedModules = (contents: string) => {
  const importedPaths: string[] = [];
  const parsedContents = parse(contents, {
    sourceType: 'unambiguous',
    strictMode: false,
    errorRecovery: true,
    plugins: ['jsx', 'typescript', 'decorators'],
  });

  traverse(parsedContents, {
    ImportDeclaration({ node }) {
      importedPaths.push(node.source.value);
    },
    ExportAllDeclaration({ node }) {
      importedPaths.push(node.source.value);
    },
    ExportNamedDeclaration({ node }) {
      if (node.source) {
        importedPaths.push(node.source.value);
      }
    },
    CallExpression({ node }) {
      if ('name' in node.callee && node.callee.name === 'require') {
        if (node.arguments.length === 1) {
          const importPathNode = node.arguments[0]!;
          if (importPathNode!.type === 'StringLiteral') {
            importedPaths.push(importPathNode.value);
          }
        }
      }
    },
  });

  return importedPaths;
};
