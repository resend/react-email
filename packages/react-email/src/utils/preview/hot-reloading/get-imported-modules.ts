import { parse } from '@babel/parser';

import traverseModule from '@babel/traverse';

const traverse =
  // we keep this check here so that this still works with the dev:preview
  // script's use of tsx
  typeof traverseModule === 'function'
    ? traverseModule
    : traverseModule.default;

export const getImportedModules = (contents: string) => {
  const importedPaths: string[] = [];
  const parsedContents = parse(contents, {
    sourceType: 'unambiguous',
    strictMode: false,
    errorRecovery: true,
    plugins: ['jsx', 'typescript', 'decorators'],
  });

  traverse(parsedContents, {
    ImportDeclaration({ node }) {
      importedPaths.push(node.source.value);
    },
    ExportAllDeclaration({ node }) {
      importedPaths.push(node.source.value);
    },
    ExportNamedDeclaration({ node }) {
      if (node.source) {
        importedPaths.push(node.source.value);
      }
    },
    TSExternalModuleReference({ node }) {
      importedPaths.push(node.expression.value);
    },
    CallExpression({ node }) {
      if ('name' in node.callee && node.callee.name === 'require') {
        if (node.arguments.length === 1) {
          const importPathNode = node.arguments[0]!;
          if (importPathNode!.type === 'StringLiteral') {
            importedPaths.push(importPathNode.value);
          }
        }
      }
    },
  });

  return importedPaths;
};
