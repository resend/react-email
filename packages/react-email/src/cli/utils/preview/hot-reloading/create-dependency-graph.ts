import { promises as fs, existsSync, statSync } from 'node:fs';
import path from 'node:path';
import type { EventName } from 'chokidar/handler';
import { isDev } from '../start-dev-server';
import { getImportedModules } from './get-imported-modules';

interface Module {
  path: string;

  dependencyPaths: string[];
  dependentPaths: string[];

  moduleDependencies: string[];
}

export type DependencyGraph = Record</* path to module */ string, Module>;

const readAllFilesInsideDirectory = async (directory: string) => {
  let allFilePaths: string[] = [];

  const topLevelDirents = await fs.readdir(directory, { withFileTypes: true });

  for await (const dirent of topLevelDirents) {
    const pathToDirent = path.join(directory, dirent.name);
    if (dirent.isDirectory()) {
      allFilePaths = allFilePaths.concat(
        await readAllFilesInsideDirectory(pathToDirent),
      );
    } else {
      allFilePaths.push(pathToDirent);
    }
  }

  return allFilePaths;
};

const isJavascriptModule = (filePath: string) => {
  const extensionName = path.extname(filePath);

  return ['.js', '.ts', '.jsx', '.tsx', '.mjs', '.cjs'].includes(extensionName);
};

const checkFileExtensionsUntilItExists = (
  pathWithoutExtension: string,
): string | undefined => {
  if (existsSync(`${pathWithoutExtension}.ts`)) {
    return `${pathWithoutExtension}.ts`;
  }

  if (existsSync(`${pathWithoutExtension}.tsx`)) {
    return `${pathWithoutExtension}.tsx`;
  }

  if (existsSync(`${pathWithoutExtension}.js`)) {
    return `${pathWithoutExtension}.js`;
  }

  if (existsSync(`${pathWithoutExtension}.jsx`)) {
    return `${pathWithoutExtension}.jsx`;
  }

  if (existsSync(`${pathWithoutExtension}.mjs`)) {
    return `${pathWithoutExtension}.mjs`;
  }

  if (existsSync(`${pathWithoutExtension}.cjs`)) {
    return `${pathWithoutExtension}.cjs`;
  }
};

/**
 * Creates a stateful dependency graph that is structured in a way that you can get
 * the dependents of a module from its path.
 *
 * Stateful in the sense that it provides a `getter` and an "`updater`". The updater
 * will receive changes to the files, that can be perceived through some file watching mechanism,
 * so that it doesn't need to recompute the entire dependency graph but only the parts changed.
 */
export const createDependencyGraph = async (directory: string) => {
  const filePaths = await readAllFilesInsideDirectory(directory);
  const modulePaths = filePaths.filter(isJavascriptModule);
  const graph: DependencyGraph = Object.fromEntries(
    modulePaths.map((path) => [
      path,
      {
        path,
        dependencyPaths: [],
        dependentPaths: [],
        moduleDependencies: [],
      },
    ]),
  );

  const getDependencyPaths = async (filePath: string) => {
    const contents = await fs.readFile(filePath, 'utf8');

    const importedPaths = getImportedModules(contents);
    const importedPathsRelativeToDirectory = importedPaths.map(
      (dependencyPath) => {
        const isModulePath = !dependencyPath.startsWith('.');

        /*
          path.isAbsolute will return false if the path looks like JavaScript module imports
          e.g. path.isAbsolute('react-dom/server') will return false, but for our purposes this
          path is not a relative one.
        */
        if (isModulePath || path.isAbsolute(dependencyPath)) {
          return dependencyPath;
        }

        let pathToDependencyFromDirectory = path.resolve(
          /*
            path.resolve resolves paths differently from what imports on javascript do.

            So if we wouldn't do this, for an email at "/path/to/email.tsx" with a dependecy path of "./other-email" 
            would end up going into /path/to/email.tsx/other-email instead of /path/to/other-email which is the
            one the import is meant to go to
          */
          path.dirname(filePath),
          dependencyPath,
        );

        let isDirectory = false;
        try {
          // will throw if the the file is not existant
          isDirectory = statSync(pathToDependencyFromDirectory).isDirectory();
        } catch (_) {}
        if (isDirectory) {
          const pathToSubDirectory = pathToDependencyFromDirectory;
          const pathWithExtension = checkFileExtensionsUntilItExists(
            `${pathToSubDirectory}/index`,
          );
          if (pathWithExtension) {
            pathToDependencyFromDirectory = pathWithExtension;
          } else if (isDev) {
            // only warn about this on development as it is probably going to be irrelevant otherwise
            console.warn(
              `Could not find index file for directory at ${pathToDependencyFromDirectory}. This is probably going to cause issues with both hot reloading and your code.`,
            );
          }
        }

        /*
          If the path to the dependency does not include a file extension, such that our check
          for it being a javascript module fails, then we can assume it has the same as the `filePath`
        */
        if (!isJavascriptModule(pathToDependencyFromDirectory)) {
          const pathWithExtension = checkFileExtensionsUntilItExists(
            pathToDependencyFromDirectory,
          );

          if (pathWithExtension) {
            pathToDependencyFromDirectory = pathWithExtension;
          } else if (isDev) {
            // only warn about this on development as it is probably going to be irrelevant otherwise
            console.warn(
              `Could not determine the file extension for the file at ${pathToDependencyFromDirectory}`,
            );
          }
        }

        return pathToDependencyFromDirectory;
      },
    );

    const moduleDependencies = importedPathsRelativeToDirectory.filter(
      (dependencyPath) =>
        !dependencyPath.startsWith('.') && !path.isAbsolute(dependencyPath),
    );

    const nonNodeModuleImportPathsRelativeToDirectory =
      importedPathsRelativeToDirectory.filter(
        (dependencyPath) =>
          dependencyPath.startsWith('.') || path.isAbsolute(dependencyPath),
      );

    return {
      dependencyPaths: nonNodeModuleImportPathsRelativeToDirectory,
      moduleDependencies,
    };
  };

  const updateModuleDependenciesInGraph = async (moduleFilePath: string) => {
    const module = graph[moduleFilePath] ?? {
      path: moduleFilePath,
      dependencyPaths: [],
      dependentPaths: [],
      moduleDependencies: [],
    };

    const { moduleDependencies, dependencyPaths: newDependencyPaths } =
      await getDependencyPaths(moduleFilePath);

    module.moduleDependencies = moduleDependencies;

    // we go through these to remove the ones that don't exist anymore
    for (const dependencyPath of module.dependencyPaths) {
      // Looping through only the ones that were on the dependencyPaths but are not
      // in the newDependencyPaths
      if (newDependencyPaths.includes(dependencyPath)) continue;

      const dependencyModule = graph[dependencyPath];
      if (dependencyModule !== undefined) {
        dependencyModule.dependentPaths =
          dependencyModule.dependentPaths.filter(
            (dependentPath) => dependentPath !== moduleFilePath,
          );
      }
    }

    module.dependencyPaths = newDependencyPaths;

    for (const dependencyPath of newDependencyPaths) {
      const dependencyModule = graph[dependencyPath];
      if (
        dependencyModule !== undefined &&
        !dependencyModule.dependentPaths.includes(moduleFilePath)
      ) {
        dependencyModule.dependentPaths.push(moduleFilePath);
      } else {
        /*
          This import path might have not been initialized as it can be outside
          of the original directory we looked into.
        */
        graph[dependencyPath] = {
          path: dependencyPath,
          moduleDependencies: [],
          dependencyPaths: [],
          dependentPaths: [moduleFilePath],
        };
      }
    }

    graph[moduleFilePath] = module;
  };

  for (const filePath of modulePaths) {
    await updateModuleDependenciesInGraph(filePath);
  }

  const removeModuleFromGraph = (filePath: string) => {
    const module = graph[filePath];
    if (module) {
      for (const dependencyPath of module.dependencyPaths) {
        if (graph[dependencyPath]) {
          graph[dependencyPath]!.dependentPaths = graph[
            dependencyPath
          ]!.dependentPaths.filter(
            (dependentPath) => dependentPath !== filePath,
          );
        }
      }
      delete graph[filePath];
    }
  };

  return [
    graph,
    async (event: EventName, pathToModified: string) => {
      switch (event) {
        case 'change':
          if (isJavascriptModule(pathToModified)) {
            await updateModuleDependenciesInGraph(pathToModified);
          }
          break;
        case 'add':
          if (isJavascriptModule(pathToModified)) {
            await updateModuleDependenciesInGraph(pathToModified);
          }
          break;
        case 'addDir': {
          const filesInsideAddedDirectory =
            await readAllFilesInsideDirectory(pathToModified);
          const modulesInsideAddedDirectory =
            filesInsideAddedDirectory.filter(isJavascriptModule);
          for await (const filePath of modulesInsideAddedDirectory) {
            await updateModuleDependenciesInGraph(filePath);
          }
          break;
        }
        case 'unlink':
          if (isJavascriptModule(pathToModified)) {
            removeModuleFromGraph(pathToModified);
          }
          break;
        case 'unlinkDir': {
          const filesInsideDeletedDirectory =
            await readAllFilesInsideDirectory(pathToModified);
          const modulesInsideDeletedDirectory =
            filesInsideDeletedDirectory.filter(isJavascriptModule);
          for await (const filePath of modulesInsideDeletedDirectory) {
            removeModuleFromGraph(filePath);
          }
          break;
        }
      }
    },
    {
      resolveDependentsOf: function resolveDependentsOf(pathToModule: string) {
        const moduleEntry = graph[pathToModule];
        const dependentPaths: Array<string> = [];

        if (moduleEntry) {
          for (const dependentPath of moduleEntry.dependentPaths) {
            const dependentsOfDependent = resolveDependentsOf(dependentPath);
            dependentPaths.push(...dependentsOfDependent);
            dependentPaths.push(dependentPath);
          }
        }

        return dependentPaths;
      },
    },
  ] as const;
};
