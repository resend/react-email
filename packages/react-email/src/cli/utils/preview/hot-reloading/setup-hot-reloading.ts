import type http from 'node:http';
import path from 'node:path';
import { watch } from 'chokidar';
import debounce from 'debounce';
import { type Socket, Server as SocketServer } from 'socket.io';
import type { HotReloadChange } from '../../../../utils/types/hot-reload-change';
import { createDependencyGraph } from './create-dependency-graph';

export const setupHotreloading = async (
  devServer: http.Server,
  emailDirRelativePath: string,
) => {
  let clients: Socket[] = [];
  const io = new SocketServer(devServer);

  io.on('connection', (client) => {
    clients.push(client);

    client.on('disconnect', () => {
      clients = clients.filter((item) => item !== client);
    });
  });

  // used to keep track of all changes
  // and send them at once to the preview app through the web socket
  let changes = [] as HotReloadChange[];

  const reload = debounce(() => {
    // we detect these using the useHotreload hook on the Next app
    clients.forEach((client) => {
      client.emit('reload', changes);
    });

    changes = [];
  }, 150);

  const absolutePathToEmailsDirectory = path.resolve(
    process.cwd(),
    emailDirRelativePath,
  );

  const [dependencyGraph, updateDependencyGraph, { resolveDependentsOf }] =
    await createDependencyGraph(absolutePathToEmailsDirectory);

  const watcher = watch('', {
    ignoreInitial: true,
    cwd: absolutePathToEmailsDirectory,
  });

  const getFilesOutsideEmailsDirectory = () =>
    Object.keys(dependencyGraph).filter((p) =>
      path.relative(absolutePathToEmailsDirectory, p).startsWith('..'),
    );
  let filesOutsideEmailsDirectory = getFilesOutsideEmailsDirectory();
  // adds in to be watched separately all of the files that are outside of
  // the user's emails directory
  for (const p of filesOutsideEmailsDirectory) {
    watcher.add(p);
  }

  const exit = async () => {
    await watcher.close();
  };
  process.on('SIGINT', exit);
  process.on('uncaughtException', exit);

  watcher.on('all', async (event, relativePathToChangeTarget) => {
    const file = relativePathToChangeTarget.split(path.sep);
    if (file.length === 0) {
      return;
    }
    const pathToChangeTarget = path.resolve(
      absolutePathToEmailsDirectory,
      relativePathToChangeTarget,
    );

    await updateDependencyGraph(event, pathToChangeTarget);

    const newFilesOutsideEmailsDirectory = getFilesOutsideEmailsDirectory();
    // updates the files outside of the user's emails directory by unwatching
    // the inexistant ones and watching the new ones
    //
    // this is necessary to avoid the issue mentioned here https://github.com/resend/react-email/issues/1433#issuecomment-2177515290
    for (const p of filesOutsideEmailsDirectory) {
      if (!newFilesOutsideEmailsDirectory.includes(p)) {
        watcher.unwatch(p);
      }
    }
    for (const p of newFilesOutsideEmailsDirectory) {
      if (!filesOutsideEmailsDirectory.includes(p)) {
        watcher.add(p);
      }
    }
    filesOutsideEmailsDirectory = newFilesOutsideEmailsDirectory;

    changes.push({
      event,
      filename: relativePathToChangeTarget,
    });

    // These dependents are dependents resolved recursively, so even dependents of dependents
    // will be notified of this change so that we ensure that things are updated in the preview.
    for (const dependentPath of resolveDependentsOf(pathToChangeTarget)) {
      changes.push({
        event: 'change' as const,
        filename: path.relative(absolutePathToEmailsDirectory, dependentPath),
      });
    }
    reload();
  });

  return watcher;
};
