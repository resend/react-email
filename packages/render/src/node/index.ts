import type { Options } from '../shared/options';
import { render } from './render';

/**
 * @deprecated use {@link render}
 */
export const renderAsync = (element: React.ReactElement, options?: Options) => {
  return render(element, options);
};

export * from '../shared/options';
export * from '../shared/utils/pretty';
export * from '../shared/utils/to-plain-text';
export * from '../shared/utils/to-resend-template';
export * from './render';
