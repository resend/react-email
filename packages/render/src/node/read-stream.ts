import { Writable } from 'node:stream';
import type {
  PipeableStream,
  ReactDOMServerReadableStream,
} from 'react-dom/server.browser';

const decoder = new TextDecoder('utf-8');

export const readStream = async (
  stream: PipeableStream | ReactDOMServerReadableStream,
) => {
  let result = '';

  if ('pipeTo' in stream) {
    // means it's a readable stream
    const writableStream = new WritableStream({
      write(chunk: BufferSource) {
        result += decoder.decode(chunk);
      },
    });
    await stream.pipeTo(writableStream);
  } else {
    const writable = new Writable({
      write(chunk: BufferSource, _encoding, callback) {
        result += decoder.decode(chunk);

        callback();
      },
    });
    stream.pipe(writable);

    await new Promise<void>((resolve, reject) => {
      writable.on('error', reject);
      writable.on('close', () => {
        resolve();
      });
    });
  }

  return result;
};
