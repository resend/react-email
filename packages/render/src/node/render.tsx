import { convert } from 'html-to-text';
import { Suspense } from 'react';
import type { Options } from '../shared/options';
import { plainTextSelectors } from '../shared/plain-text-selectors';
import { pretty } from '../shared/utils/pretty';
import { readStream } from './read-stream';

export const render = async (
  element: React.ReactElement,
  options?: Options,
) => {
  const suspendedElement = <Suspense>{element}</Suspense>;
  const reactDOMServer = await import('react-dom/server');

  let html!: string;
  if (Object.hasOwn(reactDOMServer, 'renderToReadableStream')) {
    html = await readStream(
      await reactDOMServer.renderToReadableStream(suspendedElement),
    );
  } else {
    await new Promise<void>((resolve, reject) => {
      const stream = reactDOMServer.renderToPipeableStream(suspendedElement, {
        async onAllReady() {
          html = await readStream(stream);
          resolve();
        },
        onError(error) {
          reject(error as Error);
        },
      });
    });
  }

  if (options?.plainText) {
    return convert(html, {
      selectors: plainTextSelectors,
      ...options.htmlToTextOptions,
    });
  }

  const doctype =
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">';

  const document = `${doctype}${html.replace(/<!DOCTYPE.*?>/, '')}`;

  if (options?.pretty) {
    return pretty(document);
  }

  return document;
};
