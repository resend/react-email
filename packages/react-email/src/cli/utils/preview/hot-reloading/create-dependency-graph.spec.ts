import { promises as fs, existsSync } from 'node:fs';
import path from 'node:path';
import {
  type DependencyGraph,
  createDependencyGraph,
} from './create-dependency-graph';

const pathToFileForTestingDependencyGraph = path.join(
  __dirname,
  '.for-dependency-graph.ts',
);

test('createDependencyGraph()', async () => {
  if (existsSync(pathToFileForTestingDependencyGraph)) {
    await fs.rm(pathToFileForTestingDependencyGraph);
  }

  const [dependencyGraph, updateDependencyGraph] =
    await createDependencyGraph(__dirname);

  const toAbsolute = (relativePath: string) => {
    return path.resolve(__dirname, relativePath);
  };

  const convertPathsToAbsolute = (graph: DependencyGraph) => {
    return Object.fromEntries(
      Object.entries(graph).map(([relativePath, module]) => {
        return [
          toAbsolute(relativePath),
          {
            path: toAbsolute(relativePath),
            dependencyPaths: module.dependencyPaths.map((p) => toAbsolute(p)),
            dependentPaths: module.dependentPaths.map((p) => toAbsolute(p)),
            moduleDependencies: module.moduleDependencies,
          },
        ];
      }),
    );
  };

  const initialDependencyGraph = convertPathsToAbsolute({
    'create-dependency-graph.ts': {
      path: 'create-dependency-graph.ts',
      dependencyPaths: ['../start-dev-server.ts', 'get-imported-modules.ts'],
      dependentPaths: [
        'create-dependency-graph.spec.ts',
        'setup-hot-reloading.ts',
      ],
      moduleDependencies: ['node:fs', 'node:path', 'chokidar/handler'],
    },
    'create-dependency-graph.spec.ts': {
      path: 'create-dependency-graph.spec.ts',
      dependencyPaths: ['create-dependency-graph.ts'],
      dependentPaths: [],
      moduleDependencies: ['node:fs', 'node:path'],
    },
    'get-imported-modules.ts': {
      path: 'get-imported-modules',
      dependentPaths: [
        'create-dependency-graph.ts',
        'get-imported-modules.spec.ts',
      ],
      dependencyPaths: [],
      moduleDependencies: ['@babel/traverse', '@babel/parser'],
    },
    'get-imported-modules.spec.ts': {
      path: 'get-imported-modules.spec.ts',
      dependencyPaths: ['get-imported-modules.ts'],
      dependentPaths: [],
      moduleDependencies: ['node:fs'],
    },
    'setup-hot-reloading.ts': {
      path: 'setup-hot-reloading.ts',
      dependencyPaths: [
        '../../../../utils/types/hot-reload-change.ts',
        'create-dependency-graph.ts',
      ],
      dependentPaths: [],
      moduleDependencies: [
        'node:http',
        'node:path',
        'chokidar',
        'debounce',
        'socket.io',
      ],
    },
    '../start-dev-server.ts': {
      path: '../start-dev-server.ts',
      dependencyPaths: [],
      dependentPaths: ['create-dependency-graph.ts'],
      moduleDependencies: [],
    },
    '../../../../utils/types/hot-reload-change.ts': {
      path: '../../../../utils/types/hot-reload-change.ts',
      dependencyPaths: [],
      dependentPaths: ['setup-hot-reloading.ts'],
      moduleDependencies: [],
    },
  } satisfies DependencyGraph);

  expect(
    dependencyGraph,
    'the initial value for the dependency graph should work with the directory of this testing file',
  ).toEqual(initialDependencyGraph);

  await fs.writeFile(
    pathToFileForTestingDependencyGraph,
    `
import {} from './setup-hot-reloading';
import {} from './get-imported-modules';
import {} from './create-dependency-graph.ts';
`,
    'utf8',
  );
  await updateDependencyGraph('add', pathToFileForTestingDependencyGraph);
  expect(
    dependencyGraph[pathToFileForTestingDependencyGraph],
    'added file to have proper dependency paths',
  ).toEqual({
    path: pathToFileForTestingDependencyGraph,
    dependentPaths: [],
    dependencyPaths: [
      toAbsolute('setup-hot-reloading.ts'),
      toAbsolute('get-imported-modules.ts'),
      toAbsolute('create-dependency-graph.ts'),
    ],
    moduleDependencies: [],
  } satisfies DependencyGraph[number]);
  expect(
    dependencyGraph[toAbsolute('setup-hot-reloading.ts')]?.dependentPaths,
  ).toContain(pathToFileForTestingDependencyGraph);
  expect(
    dependencyGraph[toAbsolute('get-imported-modules.ts')]?.dependentPaths,
  ).toContain(pathToFileForTestingDependencyGraph);
  expect(
    dependencyGraph[toAbsolute('create-dependency-graph.ts')]?.dependentPaths,
  ).toContain(pathToFileForTestingDependencyGraph);

  await fs.writeFile(
    pathToFileForTestingDependencyGraph,
    `
import {} from './setup-hot-reloading';
import {} from './create-dependency-graph.ts';
`,
    'utf8',
  );
  await updateDependencyGraph('change', pathToFileForTestingDependencyGraph);
  expect(
    dependencyGraph[pathToFileForTestingDependencyGraph],
    'changed file to have updated dependencyPaths',
  ).toEqual({
    path: pathToFileForTestingDependencyGraph,
    dependentPaths: [],
    dependencyPaths: [
      toAbsolute('setup-hot-reloading.ts'),
      toAbsolute('create-dependency-graph.ts'),
    ],
    moduleDependencies: [],
  } satisfies DependencyGraph[number]);
  expect(
    dependencyGraph[toAbsolute('setup-hot-reloading.ts')]?.dependentPaths,
  ).toContain(pathToFileForTestingDependencyGraph);
  expect(
    dependencyGraph[toAbsolute('get-imported-modules.ts')]?.dependentPaths,
    'when removing dependency on a file, the dependency should have its dependents updated to not have the testing file again',
  ).not.toContain(pathToFileForTestingDependencyGraph);
  expect(
    dependencyGraph[toAbsolute('create-dependency-graph.ts')]?.dependentPaths,
  ).toContain(pathToFileForTestingDependencyGraph);

  await fs.rm(pathToFileForTestingDependencyGraph);
  await updateDependencyGraph('unlink', pathToFileForTestingDependencyGraph);
  expect(dependencyGraph[pathToFileForTestingDependencyGraph]).toBeUndefined();
  expect(
    dependencyGraph[toAbsolute('setup-hot-reloading.ts')]?.dependentPaths,
    "should remove itself from dependents once it's unlinked",
  ).not.toContain(pathToFileForTestingDependencyGraph);
  expect(
    dependencyGraph[toAbsolute('get-imported-modules.ts')]?.dependentPaths,
    "should remove itself from dependents once it's unlinked",
  ).not.toContain(pathToFileForTestingDependencyGraph);
  expect(
    dependencyGraph[toAbsolute('create-dependency-graph.ts')]?.dependentPaths,
    "should remove itself from dependents once it's unlinked",
  ).not.toContain(pathToFileForTestingDependencyGraph);
});
