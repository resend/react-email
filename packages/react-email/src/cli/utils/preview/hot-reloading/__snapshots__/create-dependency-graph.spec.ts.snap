// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createDependencyGraph() > should have the right initial value for the dependency graph 1`] = `
{
  "/home/gabriel/Projects/resend/react-email/packages/react-email/src/cli/utils/preview/hot-reloading/test/dependency-graph/inner/data-to-import.json": {
    "dependencyPaths": [],
    "dependentPaths": [
      "/home/gabriel/Projects/resend/react-email/packages/react-email/src/cli/utils/preview/hot-reloading/test/dependency-graph/inner/general-importing-file.ts",
    ],
    "moduleDependencies": [],
    "path": "/home/gabriel/Projects/resend/react-email/packages/react-email/src/cli/utils/preview/hot-reloading/test/dependency-graph/inner/data-to-import.json",
  },
  "/home/gabriel/Projects/resend/react-email/packages/react-email/src/cli/utils/preview/hot-reloading/test/dependency-graph/inner/file-a.ts": {
    "dependencyPaths": [
      "/home/gabriel/Projects/resend/react-email/packages/react-email/src/cli/utils/preview/hot-reloading/test/dependency-graph/inner/file-b.ts",
    ],
    "dependentPaths": [
      "/home/gabriel/Projects/resend/react-email/packages/react-email/src/cli/utils/preview/hot-reloading/test/dependency-graph/inner/file-b.ts",
      "/home/gabriel/Projects/resend/react-email/packages/react-email/src/cli/utils/preview/hot-reloading/test/dependency-graph/inner/general-importing-file.ts",
    ],
    "moduleDependencies": [],
    "path": "/home/gabriel/Projects/resend/react-email/packages/react-email/src/cli/utils/preview/hot-reloading/test/dependency-graph/inner/file-a.ts",
  },
  "/home/gabriel/Projects/resend/react-email/packages/react-email/src/cli/utils/preview/hot-reloading/test/dependency-graph/inner/file-b.ts": {
    "dependencyPaths": [
      "/home/gabriel/Projects/resend/react-email/packages/react-email/src/cli/utils/preview/hot-reloading/test/dependency-graph/inner/file-a.ts",
    ],
    "dependentPaths": [
      "/home/gabriel/Projects/resend/react-email/packages/react-email/src/cli/utils/preview/hot-reloading/test/dependency-graph/inner/file-a.ts",
      "/home/gabriel/Projects/resend/react-email/packages/react-email/src/cli/utils/preview/hot-reloading/test/dependency-graph/inner/general-importing-file.ts",
    ],
    "moduleDependencies": [],
    "path": "/home/gabriel/Projects/resend/react-email/packages/react-email/src/cli/utils/preview/hot-reloading/test/dependency-graph/inner/file-b.ts",
  },
  "/home/gabriel/Projects/resend/react-email/packages/react-email/src/cli/utils/preview/hot-reloading/test/dependency-graph/inner/general-importing-file.ts": {
    "dependencyPaths": [
      "/home/gabriel/Projects/resend/react-email/packages/react-email/src/cli/utils/preview/hot-reloading/test/dependency-graph/inner/file-a.ts",
      "/home/gabriel/Projects/resend/react-email/packages/react-email/src/cli/utils/preview/hot-reloading/test/dependency-graph/inner/file-b.ts",
      "/home/gabriel/Projects/resend/react-email/packages/react-email/src/cli/utils/preview/hot-reloading/test/dependency-graph/inner/data-to-import.json",
      "/home/gabriel/Projects/resend/react-email/packages/react-email/src/cli/utils/preview/hot-reloading/test/dependency-graph/outer.ts",
    ],
    "dependentPaths": [],
    "moduleDependencies": [
      "node:os",
      "node:path",
    ],
    "path": "/home/gabriel/Projects/resend/react-email/packages/react-email/src/cli/utils/preview/hot-reloading/test/dependency-graph/inner/general-importing-file.ts",
  },
  "/home/gabriel/Projects/resend/react-email/packages/react-email/src/cli/utils/preview/hot-reloading/test/dependency-graph/inner/outer-dependency.ts": {
    "dependencyPaths": [],
    "dependentPaths": [
      "/home/gabriel/Projects/resend/react-email/packages/react-email/src/cli/utils/preview/hot-reloading/test/dependency-graph/outer.ts",
    ],
    "moduleDependencies": [],
    "path": "/home/gabriel/Projects/resend/react-email/packages/react-email/src/cli/utils/preview/hot-reloading/test/dependency-graph/inner/outer-dependency.ts",
  },
  "/home/gabriel/Projects/resend/react-email/packages/react-email/src/cli/utils/preview/hot-reloading/test/dependency-graph/outer.ts": {
    "dependencyPaths": [
      "/home/gabriel/Projects/resend/react-email/packages/react-email/src/cli/utils/preview/hot-reloading/test/dependency-graph/inner/outer-dependency.ts",
    ],
    "dependentPaths": [
      "/home/gabriel/Projects/resend/react-email/packages/react-email/src/cli/utils/preview/hot-reloading/test/dependency-graph/inner/general-importing-file.ts",
    ],
    "moduleDependencies": [],
    "path": "/home/gabriel/Projects/resend/react-email/packages/react-email/src/cli/utils/preview/hot-reloading/test/dependency-graph/outer.ts",
  },
}
`;
